Data/HashMap/Internal.hs

```diff
@@ -953,43 +952,53 @@ unsafeInsert k0 v0 m0 = runST (go h0 k0 v0 0 m0)
       where i = index h s
     go h k x s t@(Collision hy v)
         | h == hy   = return $! Collision h (updateOrSnocWith (\a _ -> (# a #)) k x v)
-        | otherwise = go h k x s $ BitmapIndexed (mask hy s) (A.singleton t)
+        | otherwise = two s h k x hy t
 {-# INLINABLE unsafeInsert #-}
 
--- | Create a map from two key-value pairs which hashes don't collide. To
--- enhance sharing, the second key-value pair is represented by the hash of its
--- key and a singleton HashMap pairing its key with its value.
+-- | Create a map from a key-value pair and a 'Leaf' or 'Collision' node with a
+-- different hash.
 --
--- Note: to avoid silly thunks, this function must be strict in the
--- key. See issue #232. We don't need to force the HashMap argument
--- because it's already in WHNF (having just been matched) and we
--- just put it directly in an array.
+-- It is the caller's responsibility to ensure that the HashMap argument is in
+-- WHNF.
 two :: Shift -> Hash -> k -> v -> Hash -> HashMap k v -> ST s (HashMap k v)
-two = go
+two s h1 k1 v1 = two' s h1 (Leaf h1 (L k1 v1))
```

Comment:
I think I ought to force that `Leaf`. At least I don't see anything else forcing it.

But then other parts of `two` are surprisingly lazy: In the `bp1 == bp2` case, the `st` subtree we put into the array is a thunk. I suspect this is done on purpose, in order to delay the allocation.

@treeowl Do you know more about this? I see that you've worked on this, e.g. in https://github.com/haskell-unordered-containers/unordered-containers/commit/ec84ad58e81c882dfa5b5d2bd11ff0498b48a108 and https://github.com/haskell-unordered-containers/unordered-containers/commit/71246dc217694f8c35be179dee43a360e0ef19a8.